Added Serialization to DB for cryptographer

K-Mike opened the pull request.

K-Mike committed:
Added Serialization to DB for cryptographer

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -43,6 +50,7 @@
     private static final String KEY_NAME = "name";
     private static final String KEY_ICON= "icon";
     private static final String KEY_PUBLIC_KEY= "public_key";
```
KEY_PUBLIC_KEY не нужен же уже.

K-Mike replied:
Я пока оставил это для совместимости с остальным кодом. Могу подписать  TODO
Надо?

cypt replied:
В чем проявляется необходимость совместимости? Лучше сразу выпиливать код который не является актуальным.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -122,7 +133,7 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
 
     // Adding new Contact
-    public synchronized void addContact(Contact contact) {
+    public synchronized void addContact(Contact contact) throws IOException{
```
Надо работу с ObjectOutputStream заключить в try-catch-finally. В finally не забыть сделать close.
В случае ошибки лучше кидать RuntimeException, т.к. мы не предполагаем что такое возможно.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographer.java:
```diff
@@ -17,27 +22,38 @@
+public class RSACryptographer implements Cryptographer, Externalizable {
+    private  PrivateKey privateKey;
+    private  PublicKey publicKey;
+    private  KeyFactory keyFactory;
```
keyFactory тебе нужен только в конструкторе, можно обойтись локальной переменной.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographer.java:
```diff
@@ -17,27 +22,38 @@
 
+    RSACryptographer() throws CryptographerException {
+        try {
+            keyFactory = KeyFactory.getInstance("RSA");
```
Надо сделать приватный конструктор и фабричный метод fromStream который будет создавать пустой объект и заполнять его из стрима. Чтобы небыло возможности создать криптографер без ключей.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -458,31 +473,46 @@ public synchronized void deleteOldMessages() {
     }
 
-    private static ContentValues contactToContentValues(Contact contact) {
+    private static ContentValues contactToContentValues(Contact contact) throws IOException{
```
Тут тоже лучше обработать IOException внутри метода и бросить RuntimeException в случае ошибки.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/models/Contact.java:
```diff
@@ -12,13 +15,26 @@
     private final String name;
     private final Icon icon;
     private final String publicKey;
+    public Cryptographer cryptographer;
```
public поле - не хорошо.

K-Mike replied:
Думал так удобнее будет

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/models/Contact.java:
```diff
@@ -12,13 +15,26 @@
+        this.serverId = serverId;
+        this.name = name;
+        this.icon = icon;
+        this.publicKey = publicKey;
```
publicKey уже не нужен же?

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographer.java:
```diff
@@ -134,4 +150,16 @@ private String stringify(String str) {
+
+    @Override
+    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+            privateKey = (PrivateKey)in.readObject();
```
Если есть такая возможность лучше сохранять ключи как байтовый массив чтобы уменьшить количество потребляемой памяти.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -458,31 +473,46 @@ public synchronized void deleteOldMessages() {
     }
 //
-    private static Contact cursorToContact(Cursor cursor) {
+    private static Contact cursorToContact(Cursor cursor) throws CryptographerException {
```
Давай все эти статические конвертаторы в отдельный класс - DatabaseAdapter

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -458,31 +473,46 @@ public synchronized void deleteOldMessages() {
         icon.create(cursor.getBlob(cursor.getColumnIndex("icon")));
+//        Serialize cryptographer
+        CryptographerFactory cryptographerFactory = new RSACryptographerFactory();
+        byte[] blob =  cursor.getBlob(cursor.getColumnIndex("cryptographer"));
```
У тебя есть константы для полей, надо использовать константы а не строковые значения.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/database/DBHandler.java:
```diff
@@ -210,22 +223,24 @@ public synchronized Contact getContactByServerId(String serverId) {
 
     // TODO: handle existing self contact properly
-    public synchronized void setOwnerContact(Contact contact) {
+    public synchronized void setOwnerContact(Contact contact) throws IOException,
```
После того как будешь обрабатывать IOException внутри конвертации, тут его можно будет убрать

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographer.java:
```diff
@@ -17,27 +22,38 @@
 
-            this.privateKey = kf.generatePrivate(new PKCS8EncodedKeySpec(privateKey));
-            this.publicKey = kf.generatePublic(new X509EncodedKeySpec(publicKey));
+            this.privateKey = keyFactory.generatePrivate(new PKCS8EncodedKeySpec(privateKey));
```
Может быть лучше сделать создание ключей privateKey, publicKey ленивым - перед тем как нужно что-то расшифровать или зашифровать. Тогда можно будет избавиться от CryptographerException в конструкторе что упростит работу с базой данных.

K-Mike committed:
Done:
1. Lazy initialisation of keys in cryptographer.
2. Store keys as byte[].
3. ObjectStreams are added to try catch finally

Done:
1. Lazy initialisation of keys in cryptographer.
2. Store keys as byte[].
3. ObjectStreams are added to try catch finally

Done:
1. Lazy initialisation of keys in cryptographer.
2. Store keys as byte[].
3. ObjectStreams are added to try catch finally

K-Mike committed:
Optimized a  contactToContentValues function

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographer.java:
```diff
@@ -2,48 +2,56 @@
+    private  KeyFactory keyFactory;
+
+    RSACryptographer(byte[] privateKey, byte[] publicKey) {
+//        privateKeyBytes = privateKey;
```
Комменты лучше сразу удалять.

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/RSACryptographerFactory.java:
```diff
@@ -60,10 +60,9 @@ public Cryptographer create(byte[] blob) throws CryptographerException {
-            ObjectInputStream is = new ObjectInputStream(in);
-            cryptographer = (Cryptographer)is.readObject();
-
+            ByteArrayInputStream byteArray = new ByteArrayInputStream(blob);
```
Стримы не закрываются

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/models/Contact.java:
```diff
@@ -11,14 +14,28 @@
         this.name = name;
         this.icon = icon;
         this.publicKey = publicKey;
+        RSACryptographerFactory cryptographerFactory = new RSACryptographerFactory();
```
Зачем этот объект?

cypt commented on app/src/main/java/com/streamdata/apps/cryptochat/cryptography/Cryptographer.java:
```diff
@@ -1,10 +1,13 @@
  * Interface: Cryptographer - encrypt and decrypt messages.
  */
-public interface Cryptographer {
+public interface Cryptographer extends Externalizable{
```
Пробел после Externalizable

K-Mike committed:
Merge branch 'master' into working_branch/database/2
